File: src/audio.ts
import { execFile } from "node:child_process";
import { constants } from "node:fs";
import { access, chmod, mkdtemp, readFile, rm, stat, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { isAbsolute, join } from "node:path";

const TARGET_CHUNK_SECONDS = 30;
const MIN_CHUNK_SECONDS = 15;
const MAX_CHUNK_SECONDS = 36;
const BOUNDARY_SEARCH_SECONDS = 6;
export const TRANSCRIPTION_BATCH_SIZE = 6;
export const TRANSCRIPTION_MODEL_NAME = "cohere-transcribe-03-2026-Q4_K_M";
const MAX_AUDIO_SECONDS = 10 * 60;
const MAX_DECODED_WAV_BYTES = MAX_AUDIO_SECONDS * 16_000 * 2 + 1024 * 1024;
const PROCESS_TIMEOUT_MS = 120_000;
const PROCESS_MAX_BUFFER = 16 * 1024 * 1024;
const PCM_SUBFORMAT_GUID = Buffer.from("0100000000001000800000aa00389b71", "hex");

const audioExtensions: Readonly<Record<string, string>> = {
  "audio/aac": "aac",
  "audio/flac": "flac",
  "audio/mp4": "m4a",
  "audio/mpeg": "mp3",
  "audio/ogg": "ogg",
  "audio/opus": "opus",
  "audio/wav": "wav",
  "audio/webm": "webm",
  "audio/x-caf": "caf",
  "audio/x-flac": "flac",
  "audio/x-m4a": "m4a",
};

export interface Pcm16Wav {
  pcm: Buffer;
  sampleCount: number;
  sampleRate: number;
}

export function parsePcm16Wav(bytes: Uint8Array): Pcm16Wav {
  const wav = Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (wav.length < 12 || wav.toString("ascii", 0, 4) !== "RIFF") {
    throw new Error("decoded audio was not a RIFF WAV file");
  }
  if (wav.toString("ascii", 8, 12) !== "WAVE") {
    throw new Error("decoded audio was not a WAVE file");
  }

  let format: Buffer | undefined;
  let pcm: Buffer | undefined;
  let offset = 12;
  while (offset + 8 <= wav.length) {
    const id = wav.toString("ascii", offset, offset + 4);
    const size = wav.readUInt32LE(offset + 4);
    const contentsStart = offset + 8;
    const contentsEnd = contentsStart + size;
    if (contentsEnd > wav.length) throw new Error(`decoded WAV contained a truncated ${id} chunk`);
    if (id === "fmt ") format = wav.subarray(contentsStart, contentsEnd);
    if (id === "data") pcm = wav.subarray(contentsStart, contentsEnd);
    offset = contentsEnd + (size & 1);
  }

  if (!format || format.length < 16) throw new Error("decoded WAV contained no valid fmt chunk");
  if (!pcm) throw new Error("decoded WAV contained no data chunk");
  const audioFormat = format.readUInt16LE(0);
  const channels = format.readUInt16LE(2);
  const sampleRate = format.readUInt32LE(4);
  const blockAlign = format.readUInt16LE(12);
  const bitsPerSample = format.readUInt16LE(14);
  const pcmFormat =
    audioFormat === 1 ||
    (audioFormat === 0xfffe &&
      format.length >= 40 &&
      format.readUInt16LE(16) >= 22 &&
      format.readUInt16LE(18) === 16 &&
      format.subarray(24, 40).equals(PCM_SUBFORMAT_GUID));
  if (
    !pcmFormat ||
    channels !== 1 ||
    sampleRate !== 16_000 ||
    blockAlign !== 2 ||
    bitsPerSample !== 16
  ) {
    throw new Error("decoded WAV was not mono 16 kHz 16-bit PCM");
  }
  if (pcm.length === 0) throw new Error("decoded WAV contained no audio samples");
  if (pcm.length % blockAlign !== 0) throw new Error("decoded WAV ended with a partial sample");
  return { pcm, sampleCount: pcm.length / blockAlign, sampleRate };
}

export function findQuietChunkBoundaries(wav: Pcm16Wav): number[] {
  const { pcm, sampleCount, sampleRate } = wav;
  const durationSeconds = sampleCount / sampleRate;
  const chunkCount = Math.max(
    1,
    Math.round(durationSeconds / TARGET_CHUNK_SECONDS),
    Math.ceil(durationSeconds / MAX_CHUNK_SECONDS),
  );
  if (chunkCount === 1) return [0, sampleCount];

  const frameSamples = Math.max(1, Math.round(sampleRate / 100));
  const frameCount = Math.ceil(sampleCount / frameSamples);
  const energy = new Float64Array(frameCount + 1);
  for (let frame = 0; frame < frameCount; frame++) {
    const start = frame * frameSamples;
    const end = Math.min(sampleCount, start + frameSamples);
    let sum = 0;
    for (let sample = start; sample < end; sample++) {
      const amplitude = pcm.readInt16LE(sample * 2);
      sum += amplitude * amplitude;
    }
    energy[frame + 1] = energy[frame]! + sum / (end - start);
  }

  const boundaries = [0];
  let start = 0;
  for (let chunksRemaining = chunkCount; chunksRemaining > 1; chunksRemaining--) {
    const target = start + Math.round((sampleCount - start) / chunksRemaining);
    const lower = Math.max(
      start + MIN_CHUNK_SECONDS * sampleRate,
      sampleCount - (chunksRemaining - 1) * MAX_CHUNK_SECONDS * sampleRate,
      target - BOUNDARY_SEARCH_SECONDS * sampleRate,
    );
    const upper = Math.min(
      start + MAX_CHUNK_SECONDS * sampleRate,
      sampleCount - (chunksRemaining - 1) * MIN_CHUNK_SECONDS * sampleRate,
      target + BOUNDARY_SEARCH_SECONDS * sampleRate,
    );
    const lowerFrame = Math.ceil(lower / frameSamples);
    const upperFrame = Math.floor(upper / frameSamples);
    const halfWindowFrames = 5;
    let bestFrame = lowerFrame;
    let bestEnergy = Number.POSITIVE_INFINITY;
    let bestDistance = Number.POSITIVE_INFINITY;
    for (let frame = lowerFrame; frame <= upperFrame; frame++) {
      const windowStart = Math.max(0, frame - halfWindowFrames);
      const windowEnd = Math.min(frameCount, frame + halfWindowFrames);
      const candidateEnergy =
        (energy[windowEnd]! - energy[windowStart]!) / (windowEnd - windowStart);
      const distance = Math.abs(frame * frameSamples - target);
      if (
        candidateEnergy < bestEnergy ||
        (candidateEnergy === bestEnergy && distance < bestDistance)
      ) {
        bestFrame = frame;
        bestEnergy = candidateEnergy;
        bestDistance = distance;
      }
    }
    start = bestFrame * frameSamples;
    boundaries.push(start);
  }
  boundaries.push(sampleCount);
  return boundaries;
}

function run(
  command: string,
  arguments_: readonly string[],
  signal?: AbortSignal,
): Promise<{ stderr: string; stdout: string }> {
  return new Promise((resolveRun, rejectRun) => {
    execFile(
      command,
      arguments_,
      {
        encoding: "utf8",
        maxBuffer: PROCESS_MAX_BUFFER,
        timeout: PROCESS_TIMEOUT_MS,
        ...(signal ? { signal } : {}),
      },
      (error, stdout, stderr) => {
        if (error) {
          Object.assign(error, { stderr, stdout });
          rejectRun(error);
        } else {
          resolveRun({ stderr, stdout });
        }
      },
    );
  });
}

function failureDetail(error: unknown): string {
  if (error && typeof error === "object") {
    const value = error as { message?: unknown; stderr?: unknown };
    if (typeof value.stderr === "string" && value.stderr.trim()) {
      return value.stderr.trim().slice(-2_000);
    }
    if (typeof value.message === "string" && value.message.trim()) return value.message.trim();
  }
  return String(error);
}

async function requireFile(
  path: string,
  setting: string,
  description: string,
  mode: number,
): Promise<void> {
  if (!isAbsolute(path)) throw new Error(`${setting} must be an absolute path`);
  try {
    const metadata = await stat(path);
    if (!metadata.isFile()) throw new Error("not a file");
    await access(path, mode);
  } catch {
    throw new Error(`${description} is unavailable at ${JSON.stringify(path)}; check ${setting}`);
  }
}

export async function decodeAudioToPcm(
  audio: Uint8Array,
  mimeType: string,
  filename: string,
  converterPath: string,
  signal?: AbortSignal,
): Promise<Pcm16Wav> {
  signal?.throwIfAborted();
  if (audio.byteLength === 0) throw new Error("audio upload was empty");
  const normalizedMimeType = mimeType.split(";", 1)[0]!.trim().toLowerCase();
  const extension = audioExtensions[normalizedMimeType];
  if (!extension) throw new Error(`${filename} has an unsupported audio format`);
  await requireFile(
    converterPath,
    "AFCONVERT_PATH",
    "the macOS afconvert audio converter",
    constants.X_OK,
  );

  let directory: string | undefined;
  try {
    directory = await mkdtemp(join(tmpdir(), "sharpi-transcribe-decode-"));
    await chmod(directory, 0o700);
    const inputPath = join(directory, `input.${extension}`);
    const decodedPath = join(directory, "decoded.wav");
    await writeFile(inputPath, audio, { mode: 0o600 });
    try {
      await run(
        converterPath,
        [inputPath, "-f", "WAVE", "-d", "LEI16@16000", "-c", "1", decodedPath],
        signal,
      );
    } catch (error) {
      throw new Error(`audio conversion failed: ${failureDetail(error)}`);
    }

    try {
      signal?.throwIfAborted();
      const decoded = await stat(decodedPath);
      if (decoded.size > MAX_DECODED_WAV_BYTES) {
        throw new Error("audio recording exceeds the 10 minute local transcription limit");
      }
      const wav = parsePcm16Wav(await readFile(decodedPath));
      if (wav.sampleCount > MAX_AUDIO_SECONDS * wav.sampleRate) {
        throw new Error("audio recording exceeds the 10 minute local transcription limit");
      }
      return wav;
    } catch (error) {
      throw new Error(`audio conversion produced invalid PCM WAV: ${failureDetail(error)}`);
    }
  } finally {
    if (directory) await rm(directory, { force: true, maxRetries: 3, recursive: true });
  }
}

File: internal/audio/audio.go
package audio

import (
	"bytes"
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const (
	BatchSize                 = 6
	ModelName                 = "cohere-transcribe-03-2026-Q4_K_M"
	targetChunkSeconds        = 30
	minimumChunkSeconds       = 15
	maximumChunkSeconds       = 36
	boundarySearchSeconds     = 6
	maximumAudioSeconds       = 10 * 60
	maximumDecodedWAVBytes    = maximumAudioSeconds*16_000*2 + 1024*1024
	conversionTimeout         = 120 * time.Second
	conversionMaximumLogBytes = 16 * 1024 * 1024
)

var pcmSubformatGUID = []byte{0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}

var extensions = map[string]string{
	"audio/aac":    "aac",
	"audio/flac":   "flac",
	"audio/mp4":    "m4a",
	"audio/mpeg":   "mp3",
	"audio/ogg":    "ogg",
	"audio/opus":   "opus",
	"audio/wav":    "wav",
	"audio/webm":   "webm",
	"audio/x-caf":  "caf",
	"audio/x-flac": "flac",
	"audio/x-m4a":  "m4a",
}

type PCM16WAV struct {
	PCM         []byte
	SampleCount int
	SampleRate  int
}

func ParsePCM16WAV(wav []byte) (PCM16WAV, error) {
	if len(wav) < 12 || string(wav[:4]) != "RIFF" {
		return PCM16WAV{}, errors.New("decoded audio was not a RIFF WAV file")
	}
	if string(wav[8:12]) != "WAVE" {
		return PCM16WAV{}, errors.New("decoded audio was not a WAVE file")
	}

	var format []byte
	var pcm []byte
	for offset := 12; offset+8 <= len(wav); {
		id := string(wav[offset : offset+4])
		size := int(binary.LittleEndian.Uint32(wav[offset+4 : offset+8]))
		contentsStart := offset + 8
		contentsEnd := contentsStart + size
		if contentsEnd < contentsStart || contentsEnd > len(wav) {
			return PCM16WAV{}, fmt.Errorf("decoded WAV contained a truncated %s chunk", id)
		}
		if id == "fmt " {
			format = wav[contentsStart:contentsEnd]
		}
		if id == "data" {
			pcm = wav[contentsStart:contentsEnd]
		}
		offset = contentsEnd + (size & 1)
	}

	if len(format) < 16 {
		return PCM16WAV{}, errors.New("decoded WAV contained no valid fmt chunk")
	}
	if pcm == nil {
		return PCM16WAV{}, errors.New("decoded WAV contained no data chunk")
	}
	audioFormat := binary.LittleEndian.Uint16(format[0:2])
	channels := binary.LittleEndian.Uint16(format[2:4])
	sampleRate := binary.LittleEndian.Uint32(format[4:8])
	blockAlign := binary.LittleEndian.Uint16(format[12:14])
	bitsPerSample := binary.LittleEndian.Uint16(format[14:16])
	pcmFormat := audioFormat == 1 || (audioFormat == 0xfffe &&
		len(format) >= 40 &&
		binary.LittleEndian.Uint16(format[16:18]) >= 22 &&
		binary.LittleEndian.Uint16(format[18:20]) == 16 &&
		bytes.Equal(format[24:40], pcmSubformatGUID))
	if !pcmFormat || channels != 1 || sampleRate != 16_000 || blockAlign != 2 || bitsPerSample != 16 {
		return PCM16WAV{}, errors.New("decoded WAV was not mono 16 kHz 16-bit PCM")
	}
	if len(pcm) == 0 {
		return PCM16WAV{}, errors.New("decoded WAV contained no audio samples")
	}
	if len(pcm)%int(blockAlign) != 0 {
		return PCM16WAV{}, errors.New("decoded WAV ended with a partial sample")
	}
	return PCM16WAV{PCM: pcm, SampleCount: len(pcm) / int(blockAlign), SampleRate: int(sampleRate)}, nil
}

func QuietChunkBoundaries(wav PCM16WAV) []int {
	durationSeconds := float64(wav.SampleCount) / float64(wav.SampleRate)
	chunkCount := max(1, int(durationSeconds/targetChunkSeconds+0.5), (wav.SampleCount+maximumChunkSeconds*wav.SampleRate-1)/(maximumChunkSeconds*wav.SampleRate))
	if chunkCount == 1 {
		return []int{0, wav.SampleCount}
	}

	frameSamples := max(1, wav.SampleRate/100)
	frameCount := (wav.SampleCount + frameSamples - 1) / frameSamples
	energy := make([]float64, frameCount+1)
	for frame := range frameCount {
		start := frame * frameSamples
		end := min(wav.SampleCount, start+frameSamples)
		var sum float64
		for sample := start; sample < end; sample++ {
			amplitude := int16(binary.LittleEndian.Uint16(wav.PCM[sample*2 : sample*2+2]))
			sum += float64(amplitude) * float64(amplitude)
		}
		energy[frame+1] = energy[frame] + sum/float64(end-start)
	}

	boundaries := []int{0}
	start := 0
	for chunksRemaining := chunkCount; chunksRemaining > 1; chunksRemaining-- {
		target := start + (wav.SampleCount-start+chunksRemaining/2)/chunksRemaining
		lower := max(
			start+minimumChunkSeconds*wav.SampleRate,
			wav.SampleCount-(chunksRemaining-1)*maximumChunkSeconds*wav.SampleRate,
			target-boundarySearchSeconds*wav.SampleRate,
		)
		upper := min(
			start+maximumChunkSeconds*wav.SampleRate,
			wav.SampleCount-(chunksRemaining-1)*minimumChunkSeconds*wav.SampleRate,
			target+boundarySearchSeconds*wav.SampleRate,
		)
		lowerFrame := (lower + frameSamples - 1) / frameSamples
		upperFrame := upper / frameSamples
		const halfWindowFrames = 5
		bestFrame := lowerFrame
		bestEnergy := float64(^uint64(0))
		bestDistance := int(^uint(0) >> 1)
		for frame := lowerFrame; frame <= upperFrame; frame++ {
			windowStart := max(0, frame-halfWindowFrames)
			windowEnd := min(frameCount, frame+halfWindowFrames)
			candidateEnergy := (energy[windowEnd] - energy[windowStart]) / float64(windowEnd-windowStart)
			distance := abs(frame*frameSamples - target)
			if candidateEnergy < bestEnergy || candidateEnergy == bestEnergy && distance < bestDistance {
				bestFrame = frame
				bestEnergy = candidateEnergy
				bestDistance = distance
			}
		}
		start = bestFrame * frameSamples
		boundaries = append(boundaries, start)
	}
	return append(boundaries, wav.SampleCount)
}

func Float32Chunks(wav PCM16WAV, boundaries []int) [][]float32 {
	chunks := make([][]float32, 0, len(boundaries)-1)
	for index := 0; index+1 < len(boundaries); index++ {
		start := boundaries[index]
		end := boundaries[index+1]
		chunk := make([]float32, end-start)
		for sample := start; sample < end; sample++ {
			amplitude := int16(binary.LittleEndian.Uint16(wav.PCM[sample*2 : sample*2+2]))
			chunk[sample-start] = float32(amplitude) / 32_768
		}
		chunks = append(chunks, chunk)
	}
	return chunks
}

func Decode(ctx context.Context, contents []byte, mimeType, filename, converterPath string) (PCM16WAV, error) {
	if len(contents) == 0 {
		return PCM16WAV{}, errors.New("audio upload was empty")
	}
	normalizedMimeType := strings.ToLower(strings.TrimSpace(strings.SplitN(mimeType, ";", 2)[0]))
	extension := extensions[normalizedMimeType]
	if extension == "" {
		return PCM16WAV{}, fmt.Errorf("%s has an unsupported audio format", filename)
	}
	if !filepath.IsAbs(converterPath) {
		return PCM16WAV{}, errors.New("AFCONVERT_PATH must be an absolute path")
	}
	converter, err := os.Stat(converterPath)
	if err != nil || !converter.Mode().IsRegular() || converter.Mode().Perm()&0o111 == 0 {
		return PCM16WAV{}, fmt.Errorf("the macOS afconvert audio converter is unavailable at %q; check AFCONVERT_PATH", converterPath)
	}

	directory, err := os.MkdirTemp("", "chinwag-decode-")
	if err != nil {
		return PCM16WAV{}, fmt.Errorf("create audio conversion directory: %w", err)
	}
	defer os.RemoveAll(directory)
	if err := os.Chmod(directory, 0o700); err != nil {
		return PCM16WAV{}, fmt.Errorf("secure audio conversion directory: %w", err)
	}
	inputPath := filepath.Join(directory, "input."+extension)
	decodedPath := filepath.Join(directory, "decoded.wav")
	if err := os.WriteFile(inputPath, contents, 0o600); err != nil {
		return PCM16WAV{}, fmt.Errorf("write temporary audio: %w", err)
	}

	conversionContext, cancel := context.WithTimeout(ctx, conversionTimeout)
	defer cancel()
	command := exec.CommandContext(conversionContext, converterPath, inputPath, "-f", "WAVE", "-d", "LEI16@16000", "-c", "1", decodedPath)
	var output limitedBuffer
	output.limit = conversionMaximumLogBytes
	command.Stdout = &output
	command.Stderr = &output
	if err := command.Run(); err != nil {
		if contextError := conversionContext.Err(); contextError != nil {
			return PCM16WAV{}, contextError
		}
		detail := strings.TrimSpace(output.String())
		if detail == "" {
			detail = err.Error()
		}
		if len(detail) > 2_000 {
			detail = detail[len(detail)-2_000:]
		}
		return PCM16WAV{}, fmt.Errorf("audio conversion failed: %s", detail)
	}

	metadata, err := os.Stat(decodedPath)
	if err != nil {
		return PCM16WAV{}, fmt.Errorf("audio conversion produced no WAV: %w", err)
	}
	if metadata.Size() > maximumDecodedWAVBytes {
		return PCM16WAV{}, errors.New("audio recording exceeds the 10 minute local transcription limit")
	}
	decoded, err := os.ReadFile(decodedPath)
	if err != nil {
		return PCM16WAV{}, fmt.Errorf("read converted audio: %w", err)
	}
	wav, err := ParsePCM16WAV(decoded)
	if err != nil {
		return PCM16WAV{}, fmt.Errorf("audio conversion produced invalid PCM WAV: %w", err)
	}
	if wav.SampleCount > maximumAudioSeconds*wav.SampleRate {
		return PCM16WAV{}, errors.New("audio recording exceeds the 10 minute local transcription limit")
	}
	return wav, nil
}

type limitedBuffer struct {
	bytes.Buffer
	limit int
}

func (buffer *limitedBuffer) Write(contents []byte) (int, error) {
	remaining := buffer.limit - buffer.Len()
	if remaining <= 0 {
		return 0, io.ErrShortWrite
	}
	if len(contents) > remaining {
		written, _ := buffer.Buffer.Write(contents[:remaining])
		return written, io.ErrShortWrite
	}
	return buffer.Buffer.Write(contents)
}

func abs(value int) int {
	if value < 0 {
		return -value
	}
	return value
}
